### DISCUSSION TITLE
[Tentative Idea] Introduce the Option for Formula and Cask Contributors to Use a Version of Homebrew's API JSON Generated Locally from Git Repository Contents

### DISCUSSION THREAD

**Discussion by RandomDSdevel** [Homebrew/discussions#5003] [Tap maintenance and Homebrew development]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I'm partly hesitant to suggest this since it introduces an additional level of abstraction, and thus overhead, for:  

 - primarily, frequent formula and cask contributors who simply keep '`HOMEBREW_NO_INSTALL_FROM_API set all th time instead of enabling it when needed and re-disabling it afterwards and 
<details class-"markdown">
  <summary>other, less compelling motivating use cases, </summary>
  <ul>
    <li>the (admittedly dwindling) class of users who keep '<code>HOMEBREW_NO_INSTALL_FROM_API</code>' set (along with '<code>HOMEBREW_NO_AUTO_UPDATE</code>') because they prefer to restrict Homebrew from making any network requests outside of '<code>brew update</code>' and automatic and manual invocations of '<code>brew fetch</code>' since they either:  </li>
      <ul>
         <li>lack a reliable Internet connection, </li>
         <li>otherwise have limited Internet bandwidth, </li>
         <li>prefer to minimize their Internet traffic/preserve their Internet bandwidth when they can, or </li>
         <li>would otherwise prefer that the extent of Homebrew's command API surface area that makes network requests stay minimal, either:  </li>
         <ul>
           <li>due to favoring the architectural design sensibility reasons of the principles of <a href="https://en.wikipedia.org/wiki/Single_responsibility_principle">single responsibility</a>, <a href="https://en.wikipedia.org/wiki/Principle_of_least_privilege">least privilege</a>, and <a href="https://en.wikipedia.org/wiki/Principle_of_least_astonishment">least astonishment</a> and/or 
           <li>for a small bit of network security peace-of-mind</li>
         </ul>
       </uL>
     </ul>
</details>
but recent discussion in Homebrew/brew#16237 has had me thinking about this a little bit again.  

## Suggestion Premise

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Would it ever make sense at any point, as a potential future improvement, to have an alternative provider for the JSON API aside from https://formulae.brew.sh that instead generates it locally as needed using the core formula and cask taps' locally-cloned/-installed Git repositories as its backing store?  

One potential thing to think about either in parallel or afterward:  

 - How does this new, locally-generated API JSON stay up to date and in sync with any local tap changes?  
   - Do formula and cask contributors constantly have to run '`brew generate-formula-api`' and/or '`brew generate-cask-api`' for the taps that they're working in?  
   - Or are they triggered automatically when changes are detected?  (What does this change detection?  '`brew`' itself?  A filesystem watchdog daemon available from some tap in a manner similar to external '`brew`' sub-commands?)

**Comment by SMillerDev** [comment]
> frequent formula and cask contributors who simply keep '`HOMEBREW_NO_INSTALL_FROM_API set all th time instead of enabling it when needed and re-disabling it afterwards

I personally think adding an API to bump formulae and casks is a better way. Very few version bumps have to be handled outside of `brew bump` nowadays anyway.
